src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.cs

```diff
@@ -2,33 +2,74 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+
 namespace System.Threading
 {
     /// <summary>
     /// A LIFO semaphore.
     /// Waits on this semaphore are uninterruptible.
     /// </summary>
-    internal sealed class LowLevelLifoSemaphore : IDisposable
+    internal sealed partial class LowLevelLifoSemaphore : IDisposable
     {
-        private WaitSubsystem.WaitableObject _semaphore;
+        private WaiterListEntry _waiterStackHead;
+        private LowLevelLock _waiterStackLock;
+        [ThreadStatic]
+        private static WaiterListEntry t_waitEntry;
 
-        public LowLevelLifoSemaphore(int initialSignalCount, int maximumSignalCount)
+        private void Create(int maximumSignalCount)
         {
-            _semaphore = WaitSubsystem.WaitableObject.NewSemaphore(initialSignalCount, maximumSignalCount);
+            _waiterStackHead = null;
+            _waiterStackLock = new LowLevelLock();
         }
 
-        public bool Wait(int timeoutMs)
+        public void Dispose()
         {
-            return WaitSubsystem.Wait(_semaphore, timeoutMs, false, true);
         }
 
-        public int Release(int count)
+        private bool WaitCore(int timeoutMs)
         {
-            return WaitSubsystem.ReleaseSemaphore(_semaphore, count);
+            WaiterListEntry waitEntry = t_waitEntry ?? (t_waitEntry = new WaiterListEntry());
+            waitEntry._monitor.Acquire();
+            try
+            {
+                _waiterStackLock.Acquire();
+                waitEntry._next = _waiterStackHead;
+                _waiterStackHead = waitEntry;
+                _waiterStackLock.Release();
+                return waitEntry._monitor.Wait(timeoutMs);
+            }
+            finally
+            {
+                waitEntry._monitor.Release();
+            }
         }
 
-        public void Dispose()
+        private void ReleaseCore(int count)
         {
+            while (count-- > 0)
+            {
+                _waiterStackLock.Acquire();
+                WaiterListEntry waitEntry = _waiterStackHead;
+                _waiterStackHead = waitEntry?._next;
+                _waiterStackLock.Release();
+                if (waitEntry != null)
```

Comment:
It looks like it would be possible that there is no waiting thread at the time of Release(n). That signal count needs to be accounted for such that WaitCore doesn't have to wait, otherwise WaitCore may wait when a signal is available. There can also be a race where while one thread is in the process of waiting a Release is called but the monitor.Signal does not do anything, then the waiter begins waiting.

Reply:
I thought that is already accounted for in the logic around `Counts._signalCount`, but I will recheck.

Reply:
Oh right. Could there be a race where a thread registers a waiter, before it actually waits on the monitor another thread tries to wake the waiter by signaling the monitor (this is ignored), then the waiter waits on the monitor and a signal is lost?

Reply:
Yep, it is there. Fixed it locally. Thanks.